### PR TITLE
Refine storage device management for kata-agent

### PR DESCRIPTION
**Pull request by jiangliu**
Refine storage device management for kata-agent by:
- use common code from kata-types and kata-sys-util crate
- maintain a StorageDevice object instead of a reference count for each storage device.
- introduce a manager for storage device handlers.

### CODE REVIEW

**Comment by jiangliu** [comment]
hi @wedsonaf , could you please help to review this?

**Comment by wedsonaf** [comment]
> hi @wedsonaf , could you please help to review this?

I'm happy to review. Please let me know you're done changing it.

**Comment by fidencio** [comment]
@jiangliu, the 3 failures we have are the same, and they'rereal:
```
not ok 2 Empty dir volume when FSGroup is specified with non-root container
# (from function `waitForProcess' in file /tmp/jenkins/workspace/kata-containers-2.0-ubuntu-20.04-PR-qemu-devmapper/go/src/github.com/kata-containers/tests/lib/common.bash, line 107,
#  in test file k8s-empty-dirs.bats, line 54)
#   `waitForProcess "${wait_time}" "${sleep_time}" "${cmd}"' failed
# INFO: k8s configured to use runtimeclass
# Image is up to date for sha256:a6487be0b50adf014d15508876d7be8185398e140055cc8f87c341e92b2ea98e
# pod/sharevol-kata created
# Name:         sharevol-kata
# Namespace:    default
# Priority:     0
# Node:         ubuntu20-b845f0/10.0.0.28
# Start Time:   Sun, 13 Aug 2023 17:26:13 +0000
# Labels:       <none>
# Annotations:  <none>
# Status:       Running
# IP:           10.244.0.15
# IPs:
#   IP:  10.244.0.15
# Containers:
#   mounttest-container:
#     Container ID:  containerd://759440444072d9201db3163c68767a9f1de189f461eb846d37b5ecd2e74c813c
#     Image:         k8s.gcr.io/e2e-test-images/agnhost:2.21
#     Image ID:      k8s.gcr.io/e2e-test-images/agnhost@sha256:ab055cd3d45f50b90732c14593a5bf50f210871bb4f91994c756fc22db6d922a
#     Port:          <none>
#     Host Port:     <none>
#     Args:
#       mounttest
#       --fs_type=/test-volume
#       --new_file_0660=/test-volume/test-file
#       --file_perm=/test-volume/test-file
#       --file_owner=/test-volume/test-file
#     State:          Terminated
#       Reason:       Completed
#       Exit Code:    0
#       Started:      Sun, 13 Aug 2023 17:26:15 +0000
#       Finished:     Sun, 13 Aug 2023 17:26:15 +0000
#     Ready:          False
#     Restart Count:  0
#     Environment:    <none>
#     Mounts:
#       /test-volume from emptydir-volume (rw)
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-s7c48 (ro)
#   mounttest-container-2:
#     Container ID:  containerd://5a9f4e57a7fc18a7865929afa843a130e4daef58d45ae63db9df8a0f183c842b
#     Image:         k8s.gcr.io/e2e-test-images/agnhost:2.21
#     Image ID:      k8s.gcr.io/e2e-test-images/agnhost@sha256:ab055cd3d45f50b90732c14593a5bf50f210871bb4f91994c756fc22db6d922a
#     Port:          <none>
#     Host Port:     <none>
#     Args:
#       mounttest
#       --fs_type=/test-volume-2
#       --new_file_0660=/test-volume-2/test-file
#       --file_perm=/test-volume-2/test-file
#       --file_owner=/test-volume-2/test-file
#     State:          Running
#       Started:      Sun, 13 Aug 2023 17:27:16 +0000
#     Ready:          True
#     Restart Count:  0
#     Environment:    <none>
#     Mounts:
#       /test-volume-2 from mem-emptydir-volume (rw)
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-s7c48 (ro)
# Conditions:
#   Type              Status
#   Initialized       True
#   Ready             False
#   ContainersReady   False
#   PodScheduled      True
# Volumes:
#   emptydir-volume:
#     Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
#     Medium:
#     SizeLimit:  <unset>
#   mem-emptydir-volume:
#     Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
#     Medium:     Memory
#     SizeLimit:  <unset>
#   kube-api-access-s7c48:
#     Type:                    Projected (a volume that contains injected data from multiple sources)
#     TokenExpirationSeconds:  3607
#     ConfigMapName:           kube-root-ca.crt
#     ConfigMapOptional:       <nil>
#     DownwardAPI:             true
# QoS Class:                   BestEffort
# Node-Selectors:              <none>
# Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
#                              node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
# Events:
#   Type    Reason     Age   From               Message
#   ----    ------     ----  ----               -------
#   Normal  Scheduled  91s   default-scheduler  Successfully assigned default/sharevol-kata to ubuntu20-b845f0
#   Normal  Pulled     89s   kubelet            Container image "k8s.gcr.io/e2e-test-images/agnhost:2.21" already present on machine
#   Normal  Created    89s   kubelet            Created container mounttest-container
#   Normal  Started    89s   kubelet            Started container mounttest-container
#   Normal  Pulled     89s   kubelet            Container image "k8s.gcr.io/e2e-test-images/agnhost:2.21" already present on machine
#   Normal  Created    89s   kubelet            Created container mounttest-container-2
#   Normal  Started    28s   kubelet            Started container mounttest-container-2
# pod "sharevol-kata" deleted
[run_kubernetes_tests.sh:153] ERROR: bats k8s-empty-dirs.bats
```

All the failures are happening when using devmapper as snapshotter (clh-devmapper, qemu-devmapper, and firecracker CIs).

**Comment by fidencio** [comment]
@jiangliu, sorry for the delay here, let me re-run the jobs and then I will point you to the logs.

/test-devmapper
/test-fc

**Comment by fidencio** [comment]
@jiangliu, when you get to http://jenkins.katacontainers.io/job/kata-containers-2.0-firecracker-ubuntu-PR/4815/ you can see a bunch of gz files, which will contain the logs you need.

The test case that's failing comes from here: https://github.com/kata-containers/tests/blob/de2c82824146fb200f99eab00bc99f934b87844e/integration/kubernetes/k8s-empty-dirs.bats#L42-L66

Let me know if you need any other info.

**Comment by jiangliu** [comment]
> @jiangliu, when you get to http://jenkins.katacontainers.io/job/kata-containers-2.0-firecracker-ubuntu-PR/4815/ you can see a bunch of gz files, which will contain the logs you need.
> 
> The test case that's failing comes from here: https://github.com/kata-containers/tests/blob/de2c82824146fb200f99eab00bc99f934b87844e/integration/kubernetes/k8s-empty-dirs.bats#L42-L66
> 
> Let me know if you need any other info.

@fidencio thanks for helping:)
It took me a whole week to shoot down the silly bug:(
Now it's ready for review now.

File: src/agent/src/mount.rs [hunk 1]
```diff
@@ -328,283 +390,270 @@ pub async fn update_ephemeral_mounts(
...
+
+#[async_trait::async_trait]
+impl StorageHandler for PmemHandler {
+    #[instrument]
+    async fn create_device(
+        &self,
+        storage: Storage,
+        ctx: &mut StorageContext,
+    ) -> Result<StorageDeviceObject> {
+        // Retrieve the device path from NVDIMM address.
```
Comment (quanweiZhou) [inline src/agent/src/mount.rs:623]: Suggest changing the comment to `Retrieve the device for pmem storage`.

**Review by quanweiZhou — COMMENTED** [review]
Thanks, @jiangliu few suggestions

File: src/agent/src/mount.rs [hunk 2]
```diff
@@ -131,22 +111,24 @@ lazy_static! {
...
-    DRIVER_NVDIMM_TYPE,
-    DRIVER_WATCHABLE_BIND_TYPE,
-];
-
-#[instrument]
-pub fn get_mounts() -> Result<String, std::io::Error> {
-    fs::read_to_string("/proc/mounts")
+#[rustfmt::skip]
+lazy_static! {
+    pub static ref STORAGE_HANDLERS: StorageHandlerManager<Arc<dyn StorageHandler>> = {
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/agent/src/mount.rs]: done

File: src/agent/src/storage/bind_watcher_handler.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2019 Ant Financial
```
Comment (quanweiZhou) [inline src/agent/src/storage/bind_watcher_handler.rs:1]: Suggest changing from 2019 to 2023, and checking other newly created files.

**Review by quanweiZhou — COMMENTED** [review]
one more comment.

File: src/agent/src/storage/bind_watcher_handler.rs (hunk 3 above)
Comment (jiangliu) [reply src/agent/src/storage/bind_watcher_handler.rs:1]: The code is copied from mount.rs, so I keep the original copyright.

File: src/agent/src/sandbox.rs [hunk 4]
```diff
@@ -40,6 +43,43 @@ pub const ERR_INVALID_CONTAINER_ID: &str = "Invalid container id";
...
+            inner: Arc::new(Mutex::new(StorageDeviceGeneric::new("".to_string()))),
+        }
+    }
+
+    pub fn from_device(device: StorageDeviceObject) -> Self {
+        Self { inner: device }
+    }
+
+    pub async fn ref_count(&self) -> u32 {
+        self.inner.lock().await.ref_count()
```
Comment (bergwolf) [inline src/agent/src/sandbox.rs:71]: The mutex is only used for reference counter protection. I wonder if it is better to simply use an atomic value and remove the extra `StorageState` abstraction:
```
pub struct StorageDeviceObject {
    count: AtomicU32,
    inner: Arc<dyn StorageDevice>,
}
```

Comment (jiangliu) [reply src/agent/src/sandbox.rs]: Good point, simplified.

**Review by bergwolf — APPROVED** [review]
lgtm!

**Review by quanweiZhou — APPROVED** [review]

**Review by lifupan — APPROVED** [review]

**Pull request merged**